### PR TITLE
[SDK] download and convert into well-typed entities

### LINKED ISSUE #6957: Task: EntityClient with well-typed entities in Rust  (downloading)
Task: Parsing & Serializing into/from Rust types

AC

There must be Rust types for API types
It is possible to automatically convert from untyped representation (map) into the typed representation (structure)
It is possible to map unencrypted entities to Rust types without going through decryption
<ul>
          <li>There are generated types for <code>Element</code>s from the model</li>
          <li>Generated types match &quot;value types&quot; semantics as close as possible
              (struct)
          </li>
      </ul>
      <p><strong>Notes</strong></p>
      <ul>
 <li>Depends on #7062</li>

          <li>Eventually (not for this task) we need to have some &quot;technical&quot; fields like &quot;finalIVs&quot;
              that are not in the metamodel
              <ul>
                  <li>we need to re-encrypt final fields exactly as-is so that server can
                      verify that they didn't change
                  </li>
                  <li>for that we need to use the same IVs as before or keep existing
                      encrypted fields around (that's what current Typescript code is
                      doing) but that has high memory cost
                  </li>
              </ul>
          </li>
      </ul>
  </div>

### PR DESCRIPTION
**Pull request by wec43**
Closes #6957.

### CODE REVIEW

File: tuta-sdk/rust/src/entity_client.rs [hunk 1]
```diff
@@ -160,3 +176,11 @@ impl EntityClient {
     //
     // pub async fn erase_list_element(&self, type_ref: &TypeRef, id: IdTuple) {}
 }
+
+fn has_encrypted_fields(type_model: &TypeModel) -> bool {
+    let mut has_encrypted = false;
```
Comment (charlag) [inline tuta-sdk/rust/src/entity_client.rs:181]: type model has `encrypted` field

File: tuta-sdk/rust/src/mail_facade.rs [hunk 2]
```diff
@@ -27,7 +27,7 @@ impl MailFacade {
 #[uniffi::export]
 impl MailFacade {
     /// Gets an email (an entity/instance of `Mail`) from the backend
-    pub async fn load_email_by_id_encrypted(&self, id_tuple: &IdTuple) -> Result<ParsedEntity, ApiCallError> {
+    pub async fn load_email_by_id_encrypted(&self, id_tuple: &IdTuple) -> Result<Mail, ApiCallError> {
```
Comment (charlag) [inline tuta-sdk/rust/src/mail_facade.rs:30]: it can't be `Mail` since it's encrypted

File: tuta-sdk/rust/src/entity_client.rs [hunk 3]
```diff
@@ -43,18 +50,23 @@ impl EntityClient {
 
 impl EntityClient {
     /// Gets an entity/instance of type `type_ref` from the backend
-    pub async fn load_element(
+    pub async fn load_element<'a, T: Entity + Deserialize<'a>>(
```
Comment (charlag) [inline tuta-sdk/rust/src/entity_client.rs:53]: I think we should keep the untyped `EntityClient` as is and use it from both typed one (the one you implement) and one with encryption (one for later)

File: tuta-sdk/rust/src/entity_client.rs (hunk 1 above)
Comment (wec43) [reply tuta-sdk/rust/src/entity_client.rs]: Fixed.

File: tuta-sdk/rust/src/mail_facade.rs (hunk 2 above)
Comment (wec43) [reply tuta-sdk/rust/src/mail_facade.rs]: Fixed.

File: tuta-sdk/rust/src/entity_client.rs (hunk 3 above)
Comment (wec43) [reply tuta-sdk/rust/src/entity_client.rs]: Fixed.

File: tuta-sdk/rust/src/typed_entity_client.rs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        entity_client: Arc<EntityClient>,
+        instance_mapper: Arc<InstanceMapper>,
+    ) -> Self {
+        TypedEntityClient {
+            entity_client,
+            instance_mapper,
+        }
+    }
+
+    pub async fn load_typed_element<T: Entity + Deserialize<'static>>(
```
Comment (charlag) [inline tuta-sdk/rust/src/typed_entity_client.rs:24]: I'd keep the name `load` (we should rename the other one too tbh)

File: tuta-sdk/rust/src/user_facade.rs [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+
+/// Provides high level functions to manipulate user entities via the REST API
+#[derive(uniffi::Object)]
+pub struct UserFacade {
+    entity_client: Arc<TypedEntityClient>,
+}
+
+fn user_type_ref() -> TypeRef {
+    TypeRef { app: "sys".to_owned(), type_: "User".to_owned() }
```
Comment (charlag) [inline tuta-sdk/rust/src/user_facade.rs:16]: you can call `User::type_ref()`?

File: tuta-sdk/rust/src/user_facade.rs [hunk 6]
```diff
@@ -0,0 +1,31 @@
+use std::borrow::ToOwned;
+use std::sync::Arc;
+use crate::{ApiCallError, TypeRef};
+use crate::entities::sys::User;
+use crate::entity_client::IdType;
+use crate::typed_entity_client::TypedEntityClient;
+
+
+/// Provides high level functions to manipulate user entities via the REST API
```
Comment (charlag) [inline tuta-sdk/rust/src/user_facade.rs:9]: The comment is wrong.

Currently in webapp `UserFacade` does not load anything and just holds the state. I am not sure if you intended it in the same way?

File: tuta-sdk/rust/src/lib.rs [hunk 7]
```diff
@@ -122,6 +131,10 @@ impl Sdk {
     pub fn mail_facade(&self) -> MailFacade {
         MailFacade::new(self.entity_client.clone())
     }
+
+    pub fn user_facade(&self) -> UserFacade {
```
Comment (charlag) [inline tuta-sdk/rust/src/lib.rs:135]: I don't know if we need to expose it to the users, maybe not yet?

**Review by charlag — APPROVED** [review]

File: tuta-sdk/rust/src/lib.rs (hunk 7 above)
Comment (wec43) [reply tuta-sdk/rust/src/lib.rs:135]: I'm just using this as a way for testing unencrypted entities. Is there a better one than `User`?

File: tuta-sdk/rust/src/user_facade.rs (hunk 5 above)
Comment (wec43) [reply tuta-sdk/rust/src/user_facade.rs:16]: Fxied.

File: tuta-sdk/rust/src/user_facade.rs (hunk 6 above)
Comment (wec43) [reply tuta-sdk/rust/src/user_facade.rs:9]: I changed it to a `FIXME`.

File: tuta-sdk/rust/src/typed_entity_client.rs (hunk 4 above)
Comment (wec43) [reply tuta-sdk/rust/src/typed_entity_client.rs]: Fixed.

File: tuta-sdk/rust/src/lib.rs (hunk 7 above)
Comment (charlag) [reply tuta-sdk/rust/src/lib.rs:135]: it's okay

File: tuta-sdk/rust/src/typed_entity_client.rs [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+use crate::entities::Entity;
+use crate::entity_client::{EntityClient, IdType};
+use crate::instance_mapper::InstanceMapper;
+
+pub struct TypedEntityClient {
+    entity_client: Arc<EntityClient>,
+    instance_mapper: Arc<InstanceMapper>,
+}
+
+impl TypedEntityClient {
```
Comment (charlag) [inline tuta-sdk/rust/src/typed_entity_client.rs:13]: should have a comment about what it is

File: tuta-sdk/rust/src/typed_entity_client.rs [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    ) -> Self {
+        TypedEntityClient {
+            entity_client,
+            instance_mapper,
+        }
+    }
+
+    pub async fn load<T: Entity + Deserialize<'static>>(
+        &self,
+        type_ref: &TypeRef,
```
Comment (charlag) [inline tuta-sdk/rust/src/typed_entity_client.rs:26]: we don't need to pass type_ref since we know `T` already, just do `T::type_ref`

Comment (wec43) [reply tuta-sdk/rust/src/typed_entity_client.rs]: True, fixed.

File: tuta-sdk/rust/src/typed_entity_client.rs (hunk 8 above)
Comment (wec43) [reply tuta-sdk/rust/src/typed_entity_client.rs:14]: Added a comment.

File: tuta-sdk/rust/src/entity_client.rs [hunk 10]
```diff
@@ -91,9 +84,20 @@ impl EntityClient {
         }
         let response_bytes = response.body.expect("no body");
         let response_entity = serde_json::from_slice::<RawEntity>(response_bytes.as_slice()).unwrap();
-        let parsed_entity = self.instance_mapper.parse(type_ref, response_entity)?;
+        let parsed_entity = self.json_serializer.parse(type_ref, response_entity)?;
         Ok(parsed_entity)
     }
+
+    pub fn get_type_model(&self, type_ref: &TypeRef) -> Result<&TypeModel, ApiCallError> {
+        let type_model = match self.type_model_provider.get_type_model(&type_ref.app, &type_ref.type_) {
```
Comment (charlag) [inline tuta-sdk/rust/src/entity_client.rs:92]: you don't need to change anything but FYI there's an `.ok_or()` for this case

**Review by charlag — APPROVED** [review]
please squash before merging

**Pull request merged**